File: problems/Ransom Note/canConstruct.cs
public class Solution {
    public bool CanConstruct(string ransomNote, string magazine) {
        var store = new Dictionary<char, int>();

        foreach (var item in magazine) {
            if (store.ContainsKey(item)) {
                ++store[item];
            } else {
                store.Add(item, 1);
            }
        }

        foreach (var item in ransomNote) {
            if (!store.ContainsKey(item) || 0 == store[item]) {
                return false;
            } else {
                --store[item];
            }
        }

        return true;
    }
}

File: problems/Ransom Note/canConstruct.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote.length() > magazine.length()) {
            return false;
        }

        HashMap<Character, Integer> letterFrequencies = new HashMap<>();

        for (char symbol : magazine.toCharArray()) {
            letterFrequencies.put(symbol, 1 + letterFrequencies.getOrDefault(symbol, 0));
        }

        for (char symbol : ransomNote.toCharArray()) {
            if (0 == letterFrequencies.getOrDefault(symbol, 0)) {
                return false;
            } else {
                letterFrequencies.put(symbol, letterFrequencies.get(symbol) - 1);
            }
        }

        return true;
    }
}
